mountpoint-s3-client/src/endpoint_config.rs

```diff
@@ -145,15 +150,50 @@ impl EndpointConfig {
                 .unwrap()
         };
 
-        let resolved_endpoint = endpoint_rule_engine
+        endpoint_rule_engine
             .resolve(endpoint_request_context)
-            .map_err(EndpointError::UnresolvedEndpoint)?;
-        let endpoint_uri = resolved_endpoint.get_url();
-        Uri::new_from_str(&allocator, endpoint_uri)
-            .map_err(|e| EndpointError::InvalidUri(InvalidUriError::CouldNotParse(e)))
+            .map_err(EndpointError::UnresolvedEndpoint)
     }
 }
 
+pub fn get_uri_from_endpoint(resolved_endpoint: ResolvedEndpoint) -> Result<Uri, EndpointError> {
```

Comment:
both this and `get_auth_scheme_from_endpoint` should really be methods of a new `ResolvedEndpoint` wrapper to be defined here